lib/amqp/channel.ex

```diff
@@ -5,18 +5,28 @@ defmodule AMQP.Channel do
 
   alias AMQP.{Connection, Channel}
 
-  defstruct [:conn, :pid]
-  @type t :: %Channel{conn: Connection.t(), pid: pid}
+  defstruct [:conn, :pid, :consumer_type]
+  @type t :: %Channel{conn: Connection.t(), pid: pid, consumer_type: consumer_type}
+  @type consumer_type :: :direct | :selective
+  @type opts :: [consumer_type: consumer_type, consumer_module: module, consumer_init_args: any]
 
   @doc """
   Opens a new Channel in a previously opened Connection.
+
+  ## Options
+    * `:consumer_type` - specifies the type of consumer used as callback module.
```

Comment:
Maybe this should be `:default` or `:custom` as currently `:direct` option lets you pass anything there.
It could also be more restrictive and allow only default callback module (`DirectReceiver`) implementation - which was my original intention - then this control flag had more sense. @ono what do you think?

Reply:
`:default` and `:direct` might be my preference. I am not comfortable with `:custom` sets DirectReceiver as a default consumer. `:direct` makes more sense to me. we can introduce `:custom` in future if required as an additional option.

Reply:
The other option is to eliminate this option and take `{consumer_module, args}` as a tuple like Erlang library does. Then make the following changes:

* Rename `DirectReciver` to `DirectConsumer`
* Add a proper documentation to `DirectConsumer` module

Then the user can use it simply like this:

```elixir
chan = AMQP.Channel.open(conn) # call :amqp_connection.open_channel/1
chan = AMQP.Channel.open(conn, {AMQP.DirectConsumer, self()}) # call :amqp_connection.open_channel/2
```

This approach makes more sense to me as direct or non-direct is not a right concern on channel layer. What do you think?

Reply:
Will do the latter, as it is better aligned with how Erlang library works.
Side note - maybe we would like to change `Receiver` to implement `amqp_gen_consumer` and be a substitute of `amqp_selective_receiver` as it more or less does the same thing ? As a separate PR ofc.

Reply:
The main purpose of Receiver is...

- to convert the data from Erlang library(Record) to Elixir friendly data type(Map)
- to receive messages for a channel in a single process so we can ensure the order of messages (deliver, returns etc.)

Now... I am a bit lost of the purpose of this PR :sweat:

To align 100%, could you write down an example code with direct_consumer and how it solves the problems?